motd/src/lib.rs

```diff
@@ -43,3 +255,454 @@ impl std::fmt::Display for Error {
+        //         caller has set things up since the loop means we are in-bounds.
+        let msg = unsafe { *msgs[i] };
+
+        let msg_style = pam_sys::types::PamMessageStyle::from(msg.msg_style);
+        match msg_style {
+            PamMessageStyle::PROMPT_ECHO_OFF => {
+                conv_data
+                    .errs
+                    .push(merr!("pam_motd.so asked for a password"));
+            }
+            PamMessageStyle::PROMPT_ECHO_ON => {
+                conv_data
+                    .errs
+                    .push(merr!("pam_motd.so asked for a username"));
+            }
+            PamMessageStyle::TEXT_INFO => {
+                // Safety: pam_motd.so will give us a valid cstring here.
+                let msg = unsafe { ffi::CStr::from_ptr(msg.msg) };
+                match msg.to_str() {
+                    Ok(s) => conv_data.motd_msg.push_str(s),
+                    Err(e) => conv_data
+                        .errs
+                        .push(merr!("err converting motd chunk: {}", e)),
+                }
+            }
+            PamMessageStyle::ERROR_MSG => {
+                // Safety: pam_motd.so will give us a valid cstring here.
+                let msg = unsafe { ffi::CStr::from_ptr(msg.msg) };
+                match msg.to_str() {
+                    Ok(s) => conv_data.errs.push(merr!("pam_mod.so says '{}'", s)),
+                    Err(e) => conv_data
+                        .errs
+                        .push(merr!("err converting motd err msg: {}", e)),
+                }
+            }
+        }
+    }
+
+    PamReturnCode::SUCCESS as libc::c_int
+}
+
+//
+// .so discovery logic
+//
+// TODO(ethan): I should try to resolve the location of pam_motd.so
+//       the same way that ld does. It can clearly do it much
+//       faster than the recursive walk of /usr/lib that we are doing
+//       currently.
+
+/// Specifies the strategy to use in order to find the pam_motd.so file
+/// to interrogate for the motd message.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum PamMotdResolutionStrategy {
+    /// Use the exact path to the given file. Do not attempt to do any
+    /// searching.
+    Exact(PathBuf),
+    /// Search the given list of directories. If recursive is true, also
+    /// search any subdirectories they have.
+    Search { dirs: Vec<PathBuf>, recursive: bool },
+    /// A good default. Equivalent to `Search { dirs: vec!["/usr/lib"], recursive: true }`.
+    Auto,
+}
+
+impl PamMotdResolutionStrategy {
+    fn resolve(&self) -> Result<PathBuf, Error> {
+        match self {
+            PamMotdResolutionStrategy::Exact(path) => Ok(path.to_path_buf()),
+            PamMotdResolutionStrategy::Search { dirs, recursive } => {
+                let mut err = None;
+
+                for dir in dirs.iter() {
+                    match Self::find_file(*recursive, dir, PAM_MOTD_NAME) {
```

Comment:
I wanted to avoid the dependency. I'm planning on doing a followup push to try to get the .so resolution faster, and looking at using walkdir is one option I'm planning to try. For now, I'd like to keep it with this naive solution though.

Reply:
Ack. FTR, my main consideration here is code reuse, not performance, but that should also be faster. So one way or another, please do replace the diy directory walk (followup is fine).